subcommand/webhook-cert-manager/command_test.go

```diff
@@ -0,0 +1,275 @@
+		clientset: k8s,
+	}
+	cmd.init()
+
+	file, err := ioutil.TempFile("", "config.json")
+	require.NoError(t, err)
+	defer os.Remove(file.Name())
+
+	_, err = file.Write([]byte(configFile))
+	require.NoError(t, err)
+
+	exitCh := runCommandAsynchronously(&cmd, []string{
+		"-config-file", file.Name(),
+	})
+	defer stopCommand(t, &cmd, exitCh)
+
+	timer := &retry.Timer{Timeout: 10 * time.Second, Wait: 500 * time.Millisecond}
+	retry.RunWith(timer, t, func(r *retry.R) {
+		secret1, err := k8s.CoreV1().Secrets("default").Get(context.TODO(), secretOne, metav1.GetOptions{})
+		require.NoError(r, err)
+		require.NotEqual(r, secret1.Data[v1.TLSCertKey], []byte("cert-1"))
+		require.NotEqual(r, secret1.Data[v1.TLSPrivateKeyKey], []byte("private-key-1"))
+
+		secret2, err := k8s.CoreV1().Secrets("default").Get(context.TODO(), secretTwo, metav1.GetOptions{})
+		require.NoError(r, err)
+		require.NotEqual(r, secret2.Data[v1.TLSCertKey], []byte("cert-2"))
+		require.NotEqual(r, secret2.Data[v1.TLSPrivateKeyKey], []byte("private-key-2"))
+
+		webhookConfig1, err := k8s.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Get(context.TODO(), webhookConfigOne, metav1.GetOptions{})
+		require.NoError(t, err)
+		require.NotEqual(r, webhookConfig1.Webhooks[0].ClientConfig.CABundle, caBundleOne)
+
+		webhookConfig2, err := k8s.AdmissionregistrationV1beta1().MutatingWebhookConfigurations().Get(context.TODO(), webhookConfigTwo, metav1.GetOptions{})
+		require.NoError(t, err)
+		require.NotEqual(r, webhookConfig2.Webhooks[0].ClientConfig.CABundle, caBundleTwo)
+	})
+}
+
+// This function starts the command asynchronously and returns a non-blocking chan.
+// When finished, the command will send its exit code to the channel.
+// Note that it's the responsibility of the caller to terminate the command by calling stopCommand,
+// otherwise it can run forever.
+func runCommandAsynchronously(cmd *Command, args []string) chan int {
+	// We have to run cmd.init() to ensure that the channel the command is
+	// using to watch for os interrupts is initialized. If we don't do this,
+	// then if stopCommand is called immediately, it will block forever
+	// because it calls interrupt() which will attempt to send on a nil channel.
+	cmd.init()
+	exitChan := make(chan int, 1)
+
+	go func() {
+		exitChan <- cmd.Run(args)
+	}()
+
+	return exitChan
+}
+
+func stopCommand(t *testing.T, cmd *Command, exitChan chan int) {
+	if len(exitChan) == 0 {
+		cmd.interrupt()
+	}
+	select {
+	case c := <-exitChan:
+		require.Equal(t, 0, c, string(cmd.UI.(*cli.MockUi).ErrorWriter.Bytes()))
+	}
+}
+
+const configFile = `[
+  {
+    "name": "webhookOne",
+    "tlsAutoHosts": [
+      "foo",
+      "bar",
+      "baz"
+    ],
+    "secretName": "secret-deploy-1",
+    "secretNamespace": "default"
+  },
+  {
+    "name": "webhookTwo",
+    "tlsAutoHosts": [
+      "foo-2",
+      "bar-3",
+      "baz-4"
+    ],
+    "secretName": "secret-deploy-2",
+    "secretNamespace": "default"
+  }
+]`
```

Comment:
do you think we can add a test to check that the cert bundle is updated when certs are about to expire? I think you'd probably need to figure out a way to make these configurable from tests since we're not going to wait for 24h in tests 😄

Reply:
I added a field called `certExpiry` to the command, which can override the default cert expiry if set, but it is not exposed via flags. Created a test which then sets the expiry for the test to 1 second. Let me know what you think